pandas/core/indexes/base.py

```diff
@@ -5243,16 +5247,32 @@ def get_indexer_for(self, target, **kwargs) -> np.ndarray:
 
         Returns
         -------
-        numpy.ndarray
+        numpy.ndarray[np.intp]
             List of indices.
         """
         if self._index_as_unique:
             return self.get_indexer(target, **kwargs)
         indexer, _ = self.get_indexer_non_unique(target)
         return indexer
 
+    @overload
+    def _get_indexer_non_comparable(
+        self, target: Index, method, unique: Literal[True] = ...
+    ) -> np.ndarray:
+        # returned ndarray is np.intp
+        ...
+
+    @overload
+    def _get_indexer_non_comparable(
+        self, target: Index, method, unique: Literal[False]
+    ) -> tuple[np.ndarray, np.ndarray]:
+        # both returned ndarrays are np.intp
+        ...
+
```

Comment:
try doing `reveal_type` on a file with something like
```
unique: bool

reveal_type(_get_indexer_non_comparable(unique=unique))
```

I expect it tells you that no overload variant matches it - this is why we need an overload for the `unique: bool` case (unfortunately it's not currently inferred to be the union of the `Literal[True]` and `Literal[False]` return types)